### PR TITLE
[Model WatchTower] Add ArtifactConfig (+OSS-2427) & model version management in context (partially) 

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I implemented `ArtifactConfig` and overrides for Model Objects and Deployments to be used by users in `Annotated` objects of steps outputs.

To support this feature I also extended pipeline and step runners to:
- Once steps enter context we create a new ModelVersion, if it was requested. It will be called `zenml.constants. RUNNING_MODEL_VERSION`. (see `zenml.orchestrators/step_runner.py::_prepare_model_context_for_step)
- Once step is completed we link artifacts with valid annotation to a version, which was prepared (if needed) on enter
- Once the pipeline is succeeded we `register_running_versions` giving them a valid version number (integer increments), so `running` will become `24`, if `23` was the previous greatest.

I also extended a bit structure of ArtifactLinks to support versioned links using the same name (also auto-incremented integer).

Example of attachments:
```python
@step
def a()->Tuple[Annotated[int,"1"],Annotated[int,"2",ArtifactConfig(model_name="artifact")]]:
    link_output_to_model(ArtifactConfig(model_name="manual"), output_name="1")
    return 1,2

@step
def b():
    link_output_to_model(ArtifactConfig(model_name="manual")
    return 1

@step
def c()->Annotated[int,"10"]:
    return 10

@step(enable_cache=False, model_config=ModelConfig(name="step"))
def d()->Annotated[int,"100"]:
    return 100

@pipeline(enable_cache=False, model_config=ModelConfig(name="pipe"))
def p():
    a()
    b()
    c()
    d()

if __name__=="__main__":
    p()
```
In the example above output artifacts will be linked to ModelVersions as follows:
- "2" will get Model from artifact config => artifact
- "1" is manually linked with link_output_to_model => manual
- Output of b() is not named, but is linked manually => manual
- "10" get implicit linking from pipeline => pipe
- "100" get implicit linking from step => step

So Resolution Order is: ArtifactConfig/Manual>@step>@pipeline

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by avishniakov** [comment]
For JIRA: OSS-2427

File: src/zenml/model/artifact_config.py
Comment (strickvl) [inline src/zenml/model/artifact_config.py:222]: Dumb question: what is a deployment in the context of ZenML? or what is a `DeploymentArtifact` maybe?

File: src/zenml/new/pipelines/pipeline.py
Comment (strickvl) [inline src/zenml/new/pipelines/pipeline.py:753]:
```suggestion
            KeyError: No running model version found for @step `model_config`s.
```

Maybe? Just to minimise confusion.

File: tests/integration/functional/steps/test_model_config.py
Comment (strickvl) [inline tests/integration/functional/steps/test_model_config.py:26]: 😅

**Review by strickvl — COMMENTED** [review]
Looks good AFAICS. Had one small question around the fact that people can't set both `model_version_name` and `model_stage`. Is model_stage just an enum-ified version of the version_name?

**Comment by avishniakov** [comment]
> Looks good AFAICS. Had one small question around the fact that people can't set both `model_version_name` and `model_stage`. Is model_stage just an enum-ified version of the version_name?

They can't - it is protected with `@validator("model_stage") def _validate_stage...`.
I tried to work with Union before, but pydantic is often doing some nasty conversions and making a string out of my `ModelStages`, so I can't anymore say was it a stage or someone created a version named `production`, thus I moved to a split to make it explicit. Now we don't offer custom version names, but I don't see any issue doing so (or offering other versioning like a semantic, etc.), so don't want to limit this.

File: src/zenml/model/artifact_config.py
Comment (avishniakov) [reply src/zenml/model/artifact_config.py:222]: This is `DeploymentService` in the current status quo, not sure where it will land after the "Complete Model Story" thingy. E.g. `bentoml_model_deployer_step` returns `BentoMLDeploymentService` which can be such an artifact.

Comment (strickvl) [reply src/zenml/model/artifact_config.py:222]: Maybe this was discussed already, (or maybe coming up with 'complete model story'?) but not sure if a `Service` is an artifact. Give that it's not necessarily something we can version or 'store', somehow feels like it's a separate thing? But you can probably ignore this comment as I wasn't too deep in the discussions you had on this.

**Comment by fa9r** [comment]
> but pydantic is often doing some nasty conversions and making a string out of my ModelStages

@avishniakov Pydantic is sensitive to the order of unions, e.g. if you type stuff as `Union[str, ModelStages]` it will cast everything to string, but if you type it as `Union[ModelStages, str]` it will attempt to cast it to `ModelStages` first and only fall back to string otherwise (which is the behavior you expected I guess)

**Review by strickvl — APPROVED** [review]
Approved pending tests passing and conflicts fixed.

File: src/zenml/model/artifact_config.py
Comment (avishniakov) [reply src/zenml/model/artifact_config.py:222]: No, this could be an interesting one 🙂 
What is a Service if not an Artifact - I assumed it is and here how one of them looks in current UI:
image

We will have improved UI + enrichment by metadata, but Service still is an Artifact, IMO.

Am I wrong here?

Comment (strickvl) [reply src/zenml/model/artifact_config.py:222]: Ah ok so we already have such a thing :) Then ignore my comment. I wasn't aware of that. (Though to keep in the back of your head; I'm not sure they're the same kind of thing somehow...)

File: src/zenml/model/artifact_config.py
Comment (fa9r) [reply src/zenml/model/artifact_config.py:198]: In that case, how about you add two class variables `IS_MODEL_ARTIFACT: ClassVar[bool] = False` and `IS_DEPLOYMENT_ARTIFACT: ClassVar[bool] = False` that you override in the subclasses instead, then you could simply do the following in the base class and don't need to override the method in the subclasses

```suggestion
    def link_to_model(
        self,
        artifact_uuid: UUID,
    ) -> None:
        """Link artifact to the model version.

        Args:
            artifact_uuid (UUID): The UUID of the artifact to link.
        """
        self._link_to_model_version(
            artifact_uuid,
            is_model_object=self.IS_MODEL_ARTIFACT,
            is_deployment=self.IS_DEPLOYMENT_ARTIFACT,
        )
```

File: src/zenml/new/pipelines/pipeline.py
Comment (fa9r) [reply src/zenml/new/pipelines/pipeline.py:329]: Wait, `ModelStages` is just an enum right, why would that need to be defined as a reference anyway?

Comment (fa9r) [reply src/zenml/new/pipelines/pipeline.py:329]: Some circular import stuff again? If so, I'd suggest to just move `ModelStages` to `enums.py`

**Review by fa9r — CHANGES_REQUESTED** [review]
Overall design makes sense to me now and the implementation looks good as well so far. However I think we should choose one consistent way for explicit artifact linking and remove the other one before we merge this. Probably need to wait for input from @stefannica @schustmi @htahir1 first

File: src/zenml/model/artifact_config.py
Comment (avishniakov) [reply src/zenml/model/artifact_config.py:209]: Added a warning about it

File: src/zenml/model/model_config.py
Comment (avishniakov) [reply src/zenml/model/model_config.py:93]: Yes, but during pre-release it would be nice to show what we commit for, even if it is still not working yet.

File: src/zenml/models/model_models.py
Comment (avishniakov) [reply src/zenml/models/model_models.py:262]: Regarding the while loop, I prefer to keep it - in 99% of cases I'll get results from page 1, why do I need to retrieve all versions in such a case?

File: src/zenml/orchestrators/step_runner.py
Comment (avishniakov) [reply src/zenml/orchestrators/step_runner.py:180]: This is not exactly true, cause:
```python
        try:
            model_config = get_step_context().model_config
            model_config.get_or_create_model_version()
        except StepContextError:
            return
```
`get_step_context().model_config` raises `StepContextError`if no ModelConfig is provided either in @step or in @pipeline. So I use here a "ask for excuse, not for permission" priniciple.

File: src/zenml/orchestrators/step_runner.py
Comment (avishniakov) [reply src/zenml/orchestrators/step_runner.py]: See above

File: src/zenml/model/artifact_config.py
Comment (avishniakov) [reply src/zenml/model/artifact_config.py:203]: Nice one, reworked!

File: src/zenml/models/model_base_model.py
Comment (avishniakov) [reply src/zenml/models/model_base_model.py:66]: I moved it out due to another circular import hell.

File: src/zenml/new/pipelines/pipeline.py
Comment (avishniakov) [reply src/zenml/new/pipelines/pipeline.py]: Yeah, maybe time has come - let me try moving it out to enums.

**Comment by avishniakov** [comment]
@fa9r , CI passed well for me - there were some failures, but not related to this (addressed in Discord). Can you recheck my comments and approve/reject those?

File: src/zenml/models/model_models.py
Comment (fa9r) [reply src/zenml/models/model_models.py:263]: Makes sense, fine by me 👍

File: src/zenml/orchestrators/step_runner.py
Comment (fa9r) [reply src/zenml/orchestrators/step_runner.py:183]: Ahh now I get how it works, thanks 👍

File: src/zenml/models/model_base_model.py
Comment (fa9r) [reply src/zenml/models/model_base_model.py:64]: Generally I'd suggest looking into this again to figure out whether there's a cleaner solution to the circular import, but since it's the only open issue on this PR I'd also be fine with just merging it

**Review by fa9r — APPROVED** [review]
LGTM let's get this merged 🙌

File: src/zenml/model/model_config.py
Comment (avishniakov) [reply src/zenml/model/model_config.py:93]: Removed :)

File: src/zenml/models/model_base_model.py
Comment (avishniakov) [reply src/zenml/models/model_base_model.py:64]: I tried multiple times but will proceed as is now, cause it is just a structuring thing, not an implementation problem. This small move needs too much of an effort and will touch tons of not related code.

**Pull request merged**
